### PR TITLE
Week3 Step 3 Finished

### PR DESCRIPTION
**Pull request by Xin128**
1) Copied the original pom.xml and edit the original appengine-web.xml for deploy
2) Resolve the name of "Greetings"
3) Implementation of DataServelet.java to achieve transform to Json
4) Passed the Java Linter

### CODE REVIEW

File: portfolio/pom.xml [hunk 1]
```diff
@@ -23,14 +23,29 @@
       <version>4.0.1</version>
       <scope>provided</scope>
     </dependency>
+
```
Comment (cisxingg) [inline portfolio/pom.xml]: Extra line?

File: portfolio/pom.xml [hunk 2]
```diff
@@ -23,14 +23,29 @@
...
+      <!-- Provides `mvn package appengine:run` for local testing
+           and `mvn package appengine:deploy` for deploying. -->
       <plugin>
-        <groupId>com.google.appengine</groupId>
+        <groupId>com.google.cloud.tools</groupId>
         <artifactId>appengine-maven-plugin</artifactId>
-        <version>1.9.71</version>
+        <version>2.2.0</version>
+        <configuration>
+          <!-- TODO: set project ID. -->
```
Comment (cisxingg) [inline portfolio/pom.xml]: Still a todo? Can remove if it is not the case anymore.

File: portfolio/pom.xml [hunk 3]
```diff
@@ -23,14 +23,29 @@
...
+    
+    <dependency>
+      <groupId>com.google.code.gson</groupId>
+      <artifactId>gson</artifactId>
+      <version>2.8.6</version>
+    </dependency>
   </dependencies>
 
+  
+
```
Comment (cisxingg) [inline portfolio/pom.xml]: Extra lines?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -14,18 +14,34 @@
...
 import java.io.IOException;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+import java.util.ArrayList;
+import java.util.Arrays;
 
-/** Servlet that returns some example content. */
+/** Servlet that return some example content. */
```
Comment (cisxingg) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: "returns" seems fine?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 5]
```diff
@@ -14,18 +14,34 @@
...
 public class DataServlet extends HttpServlet {
+
+  ArrayList<String> msglist = new ArrayList<String> (Arrays.asList("Hi!","Ni Hao!","Bonjour!","Hola!"));
+  
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("Hello Xin!");
+    response.setContentType("application/json;");
+    response.getWriter().println(this.convertArrayToJsonUsingGson(msglist));
```
Comment (cisxingg) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Question: Is this needed to call convertArrayToJsonUsingGson()?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 6]
```diff
@@ -14,18 +14,34 @@
...
+  }
+
+  /**
+   * Converts a ServerStats instance into a JSON string using the Gson library. Note: We first added
```
Comment (cisxingg) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Nit: Sentences that start with "Note" can start in a new line.

File: portfolio/src/main/webapp/script.js [hunk 7]
```diff
@@ -37,3 +37,16 @@ function getHellofromFetchedPractice() {
     document.getElementById('fetch-container').innerText = quote;
   });
 }
+
+/**
+ * The function is to fetch a Promise from "/data" page, convert the response message 
+ * into text and add them into the fetch-container displayed on the webpage.    
+ */
+function getDatafromJsonPractice() {
```
Comment (cisxingg) [inline portfolio/src/main/webapp/script.js]: Not a great function name :D

File: portfolio/pom.xml (hunk 2 above)
Comment (warfa-google) [inline portfolio/pom.xml]: You can remove this TODO

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 4 above)
Comment (warfa-google) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: returns

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 6 above)
Comment (warfa-google) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Whats a ServerStats instance? Doesn't this convert an arraylist?

File: portfolio/src/main/webapp/script.js [hunk 8]
```diff
@@ -37,3 +37,16 @@ function getHellofromFetchedPractice() {
...
+  fetch('/data').then(response => response.json()).then((quote) => {
+      console.log(quote.x);
```
Comment (warfa-google) [inline portfolio/src/main/webapp/script.js]: Is this code used? Should we remove it

File: portfolio/src/main/webapp/script.js (hunk 7 above)
Comment (cisxingg) [inline portfolio/src/main/webapp/script.js]: Do you mean to call this function somewhere?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 5 above)
Comment (Xin128) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: They did wrap it in the sample code, which I also don't think is quite needed. I can directly convert this to a line of code

File: portfolio/src/main/webapp/script.js (hunk 8 above)
Comment (Xin128) [reply portfolio/src/main/webapp/script.js]: This is just from step 2 practice. I just removed it.

File: portfolio/src/main/webapp/script.js (hunk 7 above)
Comment (Xin128) [reply portfolio/src/main/webapp/script.js]: This is just from step 2 practice. I just removed it.

**Review by cisxingg — APPROVED** [review]
LGTM, thanks for resolving the comments! Please wait for Warfa's approval before merging.

**Review by warfa-google — APPROVED** [review]

**Pull request merged**